packages/blunderbuss/src/datastore-lock.ts

```diff
@@ -0,0 +1,149 @@
+  private uniqueId: string;
+  private kind: string;
+  private target: string;
+  private datastore: Datastore;
+  private key: Key;
+  private lockExpiry: number;
+
+  constructor(
+    lockId: string,
+    target: string,
+    lockExpiry: number = DEFAULT_LOCK_EXPIRY
+  ) {
+    this.datastore = new Datastore();
+    this.kind = `ds-lock-${lockId}`;
+    this.target = target;
+    this.key = this.datastore.key([this.kind, this.target]);
+    this.uniqueId = uuidv4();
+    this.lockExpiry = lockExpiry;
+  }
+
+  /**
+   * Acquire the lock.
+   */
+  public async acquire(): Promise<boolean> {
+    const startTime = Date.now();
+    // Avoid multiple clients waiting for the same period of time.
+    let waitTime = BACKOFF_INITIAL_DELAY + Math.floor(Math.random() * 1000);
+    const maxDelay = BACKOFF_MAX_DELAY + Math.floor(Math.random() * 1000);
+
+    while (Date.now() - startTime < LOCK_ACQUIRE_TIMEOUT) {
+      const result = await this._acquire();
+      if (result === AcquireResult.Success) {
+        return true;
+      }
+      if (result === AcquireResult.Failure) {
+        logger.debug(
+          `Failed to acquire the lock, retrying after ${waitTime} miliseconds`
+        );
+      }
+      if (result === AcquireResult.LockActive) {
+        // The lock is active, wait and retry.
+        logger.debug(
+          `The lock is active, retrying after ${waitTime} miliseconds`
+        );
+      }
+      await sleep(waitTime);
+      waitTime = Math.min(
+        maxDelay,
+        waitTime + Math.floor(Math.random() * 2000)
+      );
+    }
+    // Timeout
+    return false;
+  }
+
+  private async _acquire(): Promise<AcquireResult> {
+    const transaction = this.datastore.transaction();
+    try {
+      await transaction.run();
+      // First get the entity
+      const [entity] = await transaction.get(this.key);
+      logger.debug(entity);
+
+      if (entity === undefined || Date.now() - entity.t > this.lockExpiry) {
+        const entity = {
+          key: this.key,
+          data: {
+            t: Date.now(),
+            uuid: this.uniqueId,
+          },
+        };
+        transaction.save(entity);
+        logger.debug(entity);
+        await transaction.commit();
+        return AcquireResult.Success;
+      } else {
+        // The lock is active.
+        await transaction.rollback();
+        return AcquireResult.LockActive;
+      }
+    } catch (err) {
+      err.message = `Error acquiring a lock: ${err.message}`;
+      logger.error(err);
+      await transaction.rollback();
+      return AcquireResult.Failure;
+    }
+  }
+
+  /**
+   * Release the lock.
+   */
+  public async release(): Promise<boolean> {
+    const transaction = this.datastore.transaction();
+    try {
+      await transaction.run();
+      const [entity] = await transaction.get(this.key);
+      if (entity === undefined || entity.uuid !== this.uniqueId) {
+        await transaction.rollback();
+        return true;
+      }
```

Comment:
This section seems wrong to me - if we attempt to release a lock, it means we think we successfully acquired it earlier. If the lock is acquired by another uuid, doesn't that mean something's gone wrong? Should this throw an error instead?

Reply:
If the lock is acquired by another uuid, this means the lock is expired and another process has obtained the lock. In this case, it should not delete the lock.

Reply:
Isn't this an error condition? It essentially means that blunderbuss didn't acquire the lock for long enough and had exited the critical section prematurely. I think it would be better to throw an error in those situations, so that the error is logged and we can possible address it by increasing the lock timeout

Reply:
Changed to throw an error.